bgpd/bgp_bmp.c

```diff
@@ -2039,22 +2060,24 @@ DEFPY(no_bmp_listener_main,
 	return CMD_SUCCESS;
 }
 
-DEFPY(bmp_connect,
-      bmp_connect_cmd,
-      "[no] bmp connect HOSTNAME port (1-65535) {min-retry (100-86400000)|max-retry (100-86400000)}",
-      NO_STR
-      BMP_STR
+DEFPY(bmp_connect, bmp_connect_cmd,
+      "[no] bmp connect HOSTNAME port (1-65535) {min-retry (100-86400000)|max-retry (100-86400000)} [source <X:X::X:X|A.B.C.D>]",
```

Comment:
The BMP code does not fix ahead of time whether it will use IPv4 or IPv6 to connect to a BMP station. It would be preferable to support giving both an IPv4 and IPv6 address simultaneously to cover both cases.

(`HOSTNAME` is resolved through DNS, and having a DNS entry with v4 + v6 provides additional redundancy options for the admin.)

Reply:
I agree with you that DNS resolution may provide IPv6 or IPv4 address, thus we have to verify the coherency (if bind and connect addresses combinaisons are not matching we will use the eventually next entry provided  by DNS) of the connect command.

but the need was driven by security considerations: be sure that the connection was coming from a specific address, this allowing acl  rules on that address. the admin are knowing their deployment  and are
rather in a deterministic approach. thus it is not preferable to offer parallel support.

Reply:
So what are you going to do when someone else adds support for parallel IPv4 + IPv6 source address?

If you want a function that limits to IPv4 or IPv6, you need to add a function that limits to IPv4 or IPv6, and not have it be a side effect of another setting.

You can put this in the same feature by making it a setting like `… <ipv4-source A.B.C.D|ipv4-disable> … <ipv6-source X:X::X:X|ipv6-disable> …` (incomplete example)

Reply:
Now we have the ability to make  one (or more) command for IPv4 and one another for IPv6
we have parallel support for IPv4 and IPv6 source address

Reply:
@eqvinox , do you mind if you can review the pull request please ?

Reply:
The proposal to use multiple connect entries with different source address it not the appropriate method, as having multiple entries in the same bmp target implies having multiple parallel sessions.

When talking about parallel support, a single BMP connect entry should either do ipv4 or ipv6 connectivity.

@eqvinox , the tests done with current libcares on main Linux distribution show that when wanting to resolve a hostname, the query will return an IPv6 address when available (while an IPV4 address could be configured but will not be returned).
The issue seems to be  at https://github.com/c-ares/c-ares/blob/cares-1_14_0/ares_gethostbyname.c#L141.

Earlier versions of libcares fix this. In the meantime, while the various distribs do not offer such service, I don't see the point in having to support multiple parallel families. The initial pull request done by @fdumontet6WIND was ok.

Reply:
> The proposal to use multiple connect entries with different source address it not the appropriate method, as having multiple entries in the same bmp target implies having multiple parallel sessions.

That's why I suggested a configuration syntax with both IPv4 and IPv6 address on the same entry.

> When talking about parallel support, a single BMP connect entry should either do ipv4 or ipv6 connectivity.

It doesn't do so right now (it will use whatever is available), it shouldn't in the future, and you are pulling a statement out of thin air here.

> @eqvinox , the tests done with current libcares on main Linux distribution show that when wanting to resolve a hostname, the query will return an IPv6 address when available (while an IPV4 address could be configured but will not be returned). The issue seems to be at [https://github.com/c-ares/c-ares/blob/cares-1_14_0/ares_gethostbyname.c#L141](https://github.com/c-ares/c-ares/blob/cares-1_14_0/ares_gethostbyname.c?rgh-link-date=2022-06-03T13%3A38%3A48Z#L141).

This is completely irrelevant because you are looking at the result of only one resolver lookup. DNS isn't frozen in time, and the results can and will change over time. What may return an IPv4 address today may return an IPv6 address tomorrow. **And this is how people do IPv4→IPv6 migrations in the real world.**

> Earlier versions of libcares fix this. In the meantime, while the various distribs do not offer such service, I don't see the point in having to support multiple parallel families. The initial pull request done by @fdumontet6WIND was ok.

NAK.

It must be possible to configure an IPv4 and IPv6 address in parallel. Anything else is not appropriate for 2022 where IPv4 addresses cost $50/ea, IPv6 push is massively ongoing, and migration paths must be made whereever possible.